### PR TITLE
Performance fixes

### PR DESCRIPTION
**Pull request by boxed**
Discussion here: https://github.com/pytest-dev/pytest/issues/2206#issuecomment-433038236

TODO:

- [x] merge https://github.com/pytest-dev/pytest/pull/4247
- [x] merge https://github.com/pytest-dev/pytest/pull/4250
- [x] rebase on features

### CODE REVIEW

File: src/_pytest/pathlib.py [hunk 1]
```diff
@@ -305,3 +305,8 @@ def fnmatch_ex(pattern, path):
     else:
         name = six.text_type(path)
     return fnmatch.fnmatch(name, pattern)
+
+
+def parts(s):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/pathlib.py]: please take a look if a `Path.parents` or `Path.parts` works out here as well as this looks like a  nice potential match

**Comment by RonnyPfannschmidt** [comment]
i believe a more critical element is actually turning the paths variable into a set, at first glance it doesn't seem to be used anywhere else, and using issubset instead of any should be a massive boost since it turns a linear search loop at the python level into a hash membership test loop on the c level

**Comment by blueyed** [comment]
> should stop at the base directory of tests (current working directory?)

That would be `rootdir`, but in this case the `argpath` could/should be used probably.

> `issubset`

Currently it checks if any of the parts are in already handled paths, which appears to be different from using `issubset`, no?
That would rather be `isdisjoint` then (https://docs.python.org/3.7/library/stdtypes.html?highlight=issubset#frozenset.isdisjoint).

**Comment by blueyed** [comment]
I'm looking this now also - @boxed, is it OK to push things here, or should I rather post diffs/patches?

**Comment by boxed** [comment]
Push away

**Comment by boxed** [comment]
I tried to find out why for 1000 files there were 2 million calls to stat but I had trouble figuring this out. Maybe a lot of micro optimizations is what's needed for that issue.

File: src/_pytest/pathlib.py (hunk 1 above)
Comment (blueyed) [reply src/_pytest/pathlib.py]: Using `p = Path(s); list(p.parents) + [p]` makes it slower.
This would only make sense if it would be a `Path` already I think.

**Comment by blueyed** [comment]
> I tried to find out why for 1000 files there were 2 million calls to stat

Some of this might also be due to `.pyc` files being considered (which I've changed by skipping `__pycache__`).

**Comment by blueyed** [comment]
Re stopping at the base dir: this appears to be the case already, no?  `parts` is "only" used for checking if it is disjoint with already collected paths, but it could be only the relevant parts then, i.e. up to rootdir - but that is likely not a big deal I would assume.

**Comment by boxed** [comment]
Re skipping __pycache__: we should explicitly skip pyc files for py2.7 too right? I did that change and it gave a modest speed boost so I skipped it while looking for bigger fish. Should have kept it in hindsight.

**Comment by boxed** [comment]
Re stopping at the base die: I mean the paths function should stop at the base path. Currently it returns "/" as the first element always which is rather silly.

**Comment by boxed** [comment]
Actually, why aren't we skipping non-py-files explicitly?

**Comment by blueyed** [comment]
> I also noticed that between 3.7.0 (which I tested my performance fix on) and master, more performance regressions have landed, resulting in my test going from 7 seconds (in 3.7 with my patch) to 9.3 seconds.

My times (using https://github.com/blueyed/dotfiles/blob/abe59a331eb0aeebccba55516ffbf885c577e8c6/usr/bin/timeit-shell),
and the script from https://github.com/pytest-dev/pytest/issues/2206#issuecomment-433006348:

- 3.6.4: 2.69
- 3.7.0: 7.18
- features branch (f466105): 7.78
- master (ccdb248): 7.96
- #4237 (initial): 4.9
- #4237 (2eab084): 3.58
- #4237 (2eab084, rebased on features): 3.58


@RonnyPfannschmidt
With #4237 (2eab084) and 1000x10 files pre-collection is not really noticeable anymore and times change from 
351.50s user 10.07s system 70% cpu 8:31.79 total
to
72.00s user 4.34s system 98% cpu 1:17.18 total
(I might have had a cold cache there initially though)

**Comment by boxed** [comment]
Cool! Can't wait to try this when I get in tomorrow at work!

**Comment by blueyed** [comment]
> Re skipping pycache: we should explicitly skip pyc files for py2.7 too right?

Yes, although IIRC it was easier / better to just check the dir upfront, but could also be done in a special way for py27, of course.

> Re stopping at the base die: I mean the paths function should stop at the base path. Currently it returns "/" as the first element always which is rather silly.

You mean the "parts" function, right?
This is only string manipulation currently, and it might be more overhead to handle stopping after all, but have not looked to closely.  AFAIK this should not cause any disk I/O though.

> Actually, why aren't we skipping non-py-files explicitly?

`pytest_collect_file` handles this (via `python_files` setting) currently.  I guess there are valid use cases for non-.py files (e.g. doctests).

**Comment by boxed** [comment]
Re paths function: yep. Could be make fast with just s[len(basepath):] before splitting shouldn't add too much complexity either. Maybe a bit premature though.

**Comment by boxed** [comment]
Yep, this brings time almost back to the 3.4 level on my machine at work for the example script.

Oooh, but it brings the time down to 4.5 on the test at work, where 6s is the benchmark (although 3.6.0 was 3s). This is good stuff!

**Comment by boxed** [comment]
The test suite at TriOptima now has 104k calls to posix.stat (compared to millions before). Huge improvement! But it's still 0.8 seconds just for all the posix.stat calls. 2906 files and folders (after I delete all __pychache__ dirs and not counting things excluded by norecusedirs). That's 104273/2906 ~= 35.88 posix.stat _per file_. There are 1767 .py files, so it's 59.0 stat calls per file if you just count that.

**Comment by RonnyPfannschmidt** [comment]
@boxed lovely stats - i believe we may need to create a little trace helper to trace the stats/assign them context - i believe the import system of python generates __dozens__ of them for trying different locations in sys.path for example - i believe there is room for improvement there but its not yet clear how to materialize it yet

right now we cant even clearly measure ^^

**Comment by boxed** [comment]
I guess some of those stat calls are for python to just check if it needs to rebuild a pyc file. I made a small script that basically just imports the entire app and I get 11586 stat calls for that, so that doesn't change the math much.

**Comment by blueyed** [comment]
Great!
I assume https://github.com/pytest-dev/pytest/pull/4224 also has an impact (just merged).
I want to use lru_cache there also (https://github.com/pytest-dev/pytest/commit/1acb18b8fbdf4a06577cacf4030d5725dd57aaf4 - will create a new PR) and then we can rebase this one already.

**Comment by boxed** [comment]
My method to find out where the excessive stat calls (285054 when I started) comes from was to monkey patch it and write to a log file:

```
os._real_stat = os.stat

def monkey_stat(path, *, dir_fd=None, follow_symlinks=True):
    import traceback

    os.stat = os._real_stat

    stack = traceback.format_stack()
    with open('stack_traces.log', 'a') as f:
        f.write(''.join(stack).replace('\n', '@@@@@'))
        f.write('\n')
    os.stat = monkey_stat

    return os._real_stat(path, dir_fd=dir_fd, follow_symlinks=follow_symlinks)

os.stat = monkey_stat
```

this created a file 269115 lines long. I then fed it to Counter:


```python
stacks = list(open('stack_traces.log').readlines())
from collections import Counter
c = Counter(stacks)  
```

and the top groups have these counts:

75000
75000
34000
30000
30000
3000

So it seems fixing the problem should be fairly simple to fix by addressing just five places in the code.

To print a stack trace: `print(c.most_common()[0][0].replace('@@@@@', '\n'))`. This is the first:

```
[...snip...]
  File "/Users/andersh/Projects/pytest/src/_pytest/fixtures.py", line 1296, in parsefactories
    holderobj = node_or_obj.obj
  File "/Users/andersh/Projects/pytest/src/_pytest/python.py", line 247, in fget
    self._obj = obj = self._getobj()
  File "/Users/andersh/Projects/pytest/src/_pytest/python.py", line 439, in _getobj
    return self._importtestmodule()
  File "/Users/andersh/Projects/pytest/src/_pytest/python.py", line 449, in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
  File "/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 951, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 890, in _find_spec
  File "<frozen importlib._bootstrap>", line 867, in _find_spec_legacy
  File "<frozen importlib._bootstrap>", line 449, in spec_from_loader
  File "/Users/andersh/Projects/pytest/src/_pytest/assertion/rewrite.py", line 303, in is_package
    fd, fn, desc = self._imp_find_module(name)
  File "/Users/andersh/Projects/pytest/src/_pytest/assertion/rewrite.py", line 78, in _imp_find_module
    return imp.find_module(name, path)
  File "/Users/andersh/Downloads/venv/lib/python3.6/imp.py", line 291, in find_module
    if os.path.isfile(file_path):
  File "/Users/andersh/Downloads/venv/bin/../lib/python3.6/genericpath.py", line 30, in isfile
    st = os.stat(path)
  File "/Users/andersh/Projects/pytest/src/_pytest/main.py", line 19, in monkey_stat
    stack = traceback.format_stack()
```


The second:

```
[...snip...]
  File "/Users/andersh/Projects/pytest/src/_pytest/python.py", line 449, in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
  File "/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/Users/andersh/Projects/pytest/src/_pytest/assertion/rewrite.py", line 293, in load_module
    mod.__spec__ = spec_from_file_location(name, co.co_filename, loader=self)
  File "<frozen importlib._bootstrap_external>", line 574, in spec_from_file_location
  File "/Users/andersh/Projects/pytest/src/_pytest/assertion/rewrite.py", line 303, in is_package
    fd, fn, desc = self._imp_find_module(name)
  File "/Users/andersh/Projects/pytest/src/_pytest/assertion/rewrite.py", line 78, in _imp_find_module
    return imp.find_module(name, path)
  File "/Users/andersh/Downloads/venv/lib/python3.6/imp.py", line 291, in find_module
    if os.path.isfile(file_path):
  File "/Users/andersh/Downloads/venv/bin/../lib/python3.6/genericpath.py", line 30, in isfile
    st = os.stat(path)
  File "/Users/andersh/Projects/pytest/src/_pytest/main.py", line 19, in monkey_stat
    stack = traceback.format_stack()
```

I put @functools.lru_cache(1000) on AssertionRewritingHook.is_package, and the number of stats calls went to 173054, a reduction of 96061.

@functools.lru_cache(1000) on PytestPluginManager._getconftestmodules gets me to 132396, a further reduction of 40658.


Here my spidey sense is tingling. I have memoized all the calls, why do I still have high number of stats calls? I update my monkey patch to stat to log the paths it looks at:

```python
os._real_stat = os.stat

def monkey_stat(path, *, dir_fd=None, follow_symlinks=True):
    import traceback

    os.stat = os._real_stat

    stack = traceback.format_stack()
    with open('stack_traces.log', 'a') as f:
        f.write(''.join(stack).replace('\n', '@@@@@'))
        f.write('\n')

    with open('stat_paths.log', 'a') as f:
        f.write(path)
        f.write('\n')
    os.stat = monkey_stat

    return os._real_stat(path, dir_fd=dir_fd, follow_symlinks=follow_symlinks)

os.stat = monkey_stat
```


`cat stat_paths.log | grep "foo_99[/.]test_bar_9."` give me:

```
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py/bin
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9.so
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9.py
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/bin/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python36.zip/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/lib-dynload/foo_99.test_bar_9.pyc
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9/__init__.py
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9/__init__.pyc
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9.cpython-36m-darwin.so
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9.abi3.so
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9.so
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9.py
/usr/local/Cellar/python@3.6.5/3.6.5_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.pyc
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9/__init__.py
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9.abi3.so
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9.so
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9.py
/Users/andersh/Projects/pytest/src/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/pluggy-0.8.0-py3.6.egg/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/atomicwrites-1.2.1-py3.6.egg/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/more_itertools-4.3.0-py3.6.egg/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/attrs-18.2.0-py3.6.egg/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/six-1.11.0-py3.6.egg/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/py-1.7.0-py3.6.egg/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9/__init__.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9/__init__.pyc
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.cpython-36m-darwin.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.abi3.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.so
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.py
/Users/andersh/Downloads/venv/lib/python3.6/site-packages/foo_99.test_bar_9.pyc
/Users/andersh/Downloads/test_pytest_collect_speed/tests/foo_99/test_bar_9.py
```


These come (almost all) from imp.find_module, and if I do `--assert=plain` they go away. At this point I and @martenbr spent a lot of time trying to figure out the import hook enough to be able to stop it from doing this, but failed. 

I tried removing the lru_cache on AssertionRewritingHook.is_package and the performance didn't suffer significantly after all, so after all this work I have just one lru_cache to show for it. 

Summary: someone who understands and doesn't fear the import hook needs to look at this, I'm giving up at this point.

**Comment by boxed** [comment]
I think this should be merged now, unless someone else wants to give the assert rewriting import hook a go right now.

File: src/_pytest/config/__init__.py [hunk 2]
```diff
@@ -383,6 +384,7 @@ def _try_load_conftest(self, anchor):
                 if x.check(dir=1):
                     self._getconftestmodules(x)
 
+    @functools.lru_cache()
     def _getconftestmodules(self, path):
         if self._noconftest:
             return []
```
Comment (blueyed) [inline src/_pytest/config/__init__.py]: This will be done via https://github.com/pytest-dev/pytest/commit/1acb18b8fbdf4a06577cacf4030d5725dd57aaf4 already, please take it out here.

Comment (boxed) [reply src/_pytest/config/__init__.py]: Oh, I forgot about that. Will do.

**Comment by blueyed** [comment]
Thanks for the additional investigation.
I think we should land https://github.com/pytest-dev/pytest/pull/4247, and then rebase this on features - after all I think there are changes that might cause regressions (i.e. the `isdisjoint` part), and it should not go into a bugfix release (i.e. master).
My commits should also be either squashed or reworded - I think it is good to have them separate for git-bisecting later, but could need rewording probably.

File: src/_pytest/config/__init__.py (hunk 2 above)
Comment (boxed) [reply src/_pytest/config/__init__.py]: Hmm.. what's with max_size=None though? Seems a bit excessive?

**Comment by boxed** [comment]
Will you do the rebasing and orchestration of your PR with this? I'm not really sure what you're talking about because I don't understand the release procedure and branch strategy of pytest, so would like someone else to do that :P

**Comment by blueyed** [comment]
With this rebased on #4247 I am getting 2.8s for the test case, which is almost as good as 3.6.4 already (2.69s)! \o/

**Comment by blueyed** [comment]
@boxed 
Sure, can do it.
Basically master is for bug fixes, and features for features/behavior changes.  And while this is a performance bugfix, it might cause a regression (i.e. different collection) that should not go into a bugfix release.

Do you want to improve the parts function as mentioned in https://github.com/pytest-dev/pytest/pull/4237#issuecomment-433130322 still?

**Comment by boxed** [comment]
Not at this time. But it might be worth a look if someone can get to the bottom of the assert rewrite import hook performance problems.

File: src/_pytest/config/__init__.py (hunk 2 above)
Comment (blueyed) [reply src/_pytest/config/__init__.py]: > Hmm.. what's with max_size=None though? Seems a bit excessive?

The default is 128, which is a bit small IMHO, and I thought it should be used for everything - just like it was done with the previous manual caching.
If limiting this is should be probably in the range of 10k, but then it might not help with bigger test suites etc.

Comment (boxed) [reply src/_pytest/config/__init__.py]: I have a big test suite at work with ~5k tests that takes 25s to collect without that lru_cache and 14s with. But there is no discernible difference between infinite cache size and 128. Maybe just because the import hook is so terribly slow in comparison, but wanted to give this perspective. We might be trading more memory for no speed improvement...

Comment (boxed) [reply src/_pytest/config/__init__.py]: (and as an aside pytest 3.6.0 collects that suite in 9s)

Comment (blueyed) [reply src/_pytest/config/__init__.py]: Ah, I see.
It is likely that the cache is only used for a short period (i.e. during collection of files in a directory etc).
A limit makes sense then indeed.

Comment (blueyed) [reply src/_pytest/config/__init__.py]: Amended, 371ed5b.

#4249 might turn out to be useful there in the future.

Comment (blueyed) [reply src/_pytest/config/__init__.py]: Amended further - 94a6396, a single lru_cache appears to be enough here.

**Comment by nicoddemus** [comment]
After the PRs related to performance are merged, I think we are in good shape for a 3.10 release. 👍

File: src/_pytest/main.py [hunk 3]
```diff
@@ -469,7 +470,7 @@ def _perform_collect(self, args, genitems):
             return items
 
     def collect(self):
-        for parts in self._initialparts:
+        for parts in self._initialparts:  # noqa: F402
```
Comment (blueyed) [inline src/_pytest/main.py]: I do not really like this (it is needed because of the imported path being shadowed).
Suggestions?

**Comment by boxed** [comment]
@blueyed So the performance on features is ok with this patch? I am confused by your last comment :P

File: src/_pytest/main.py (hunk 3 above)
Comment (boxed) [reply src/_pytest/main.py]: for initialpart in self._initialparts?

File: src/_pytest/pathlib.py (hunk 1 above)
Comment (boxed) [reply src/_pytest/pathlib.py]: No surprise there really. Path.parents is a pretty complex chain of stuff that is hard to follow. Would be amazing if it wasn't slower than the simple string operations in this function.

**Comment by boxed** [comment]
Looking a bit at the performance again, using my test script with many dirs again and just running the collect, I get ~3.8s (with the profiler on), and 0.1 of that is spent in normpath (22k calls!) which seems weird to me. Looking at posixpath.py, normpath looks like this:

```python
def normpath(path):
    """Normalize path, eliminating double slashes, etc."""
    path = os.fspath(path)
    if isinstance(path, bytes):
        sep = b'/'
        empty = b''
        dot = b'.'
        dotdot = b'..'
    else:
        sep = '/'
        empty = ''
        dot = '.'
        dotdot = '..'
    if path == empty:
        return dot
    initial_slashes = path.startswith(sep)
    # POSIX allows one or two initial slashes, but treats three or more
    # as single slash.
    if (initial_slashes and
        path.startswith(sep*2) and not path.startswith(sep*3)):
        initial_slashes = 2
    comps = path.split(sep)
    new_comps = []
    for comp in comps:
        if comp in (empty, dot):
            continue
        if (comp != dotdot or (not initial_slashes and not new_comps) or
             (new_comps and new_comps[-1] == dotdot)):
            new_comps.append(comp)
        elif new_comps:
            new_comps.pop()
    comps = new_comps
    path = sep.join(comps)
    if initial_slashes:
        path = sep*initial_slashes + path
    return path or dot
```

seems to be doing an awful lot! Trying it in ipython:

```
In [10]: %timeit normpath('/foo/bar/baz')                                                                                     
2.84 µs ± 49.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

hmm... I try this:

```
In [11]: %paste 
def fast_normpath(path):
    if '.' in path or '//' in path:
        return normpath(path)
    else:
        return path
In [12]: %timeit fast_normpath('/foo/bar/baz')                                                                                
177 ns ± 0.635 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

So this simple optimization seems to make it 83 times faster in the (normal?) case of the paths not needing normalization. It does slow down the normalizing case:

```
In [15]: %timeit normpath('/foo/bar/baz//asd/../123')                                                                         
4 µs ± 59.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [16]: %timeit fast_normpath('/foo/bar/baz//asd/../123')                                                                    
4.3 µs ± 85.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

but not by much. Is this stuff worth sending a PR to CPython you think?

**Comment by blueyed** [comment]
> So the performance on features is ok with this patch? I am confused by your last comment :P

Fixed the formatting.
It seemed like (initial) times were slower compared to when we started working on it (and that might still be the case).  I've tried to see if there was some performance regression inbetween on features, but could not find it.
Initially this was also done against master (which I forgot about), so that might have added to my confusion.

> Is this stuff worth sending a PR to CPython you think?

Sure.
OTOH 0.1 out of 3.8 is not very much for us, and 22k calls is really too much for the 1.3k files in the test case IMHO.

I've removed the `noqa`, and think we're good to merge this already.
normpath and other optimizations can be done as a followup.

**Comment by nicoddemus** [comment]
Hmm AppVeyor is having trouble to clone the repository...

**Comment by blueyed** [comment]
Re-started.. likely due to GitHub issues.

**Comment by boxed** [comment]
What does AppVeyor do that travis does not?

**Comment by blueyed** [comment]
Nothing likely, but it is usually behind, i.e. falls into other time windows.
Works by now: https://ci.appveyor.com/project/pytestbot/pytest/builds/20019105

**Comment by boxed** [comment]
🎉
